<a href="https://colab.research.google.com/github/Mirko6/part-II-project/blob/main/pyKT_evals/simple_SAINT_and_DKVMN_on_ASSISTment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Done based on the [quick start tutorial](https://pykt-toolkit.readthedocs.io/en/latest/quick_start.html).

## pyKT install

In [ ]:
%pip install -U pykt-toolkit

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.3/114.3 kB 4.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 49.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 201.7/201.7 kB 2.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 184.3/184.3 kB 14.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 4.1 MB/s eta 0:00:00
  Created wheel for pathtools: filename=pathtools-0.1.2-py3-none-any.whl size=8807 sha256=96237281f16c7795b1932c9b81c1ac935a50233aa2ff8181ea8fc9bc813132b0
  Stored in directory: /root/.cache/pip/wheels/e7/f3/22/152153d6eb222ee7a56ff8617d80ee5207207a8c00a7aab794
Successfully built pathtools


In [ ]:
!git clone https://github.com/pykt-team/pykt-toolkit.git

Cloning into 'pykt-toolkit'...
remote: Enumerating objects: 6327, done.
remote: Counting objects: 100% (735/735), done.
remote: Compressing objects: 100% (308/308), done.
remote: Total 6327 (delta 528), reused 595 (delta 426), pack-reused 5592
Receiving objects: 100% (6327/6327), 33.81 MiB | 24.60 MiB/s, done.
Resolving deltas: 100% (4909/4909), done.


## assist 2009

In [ ]:
#assistment 2009 - download
assistment_2009_download_id = '1NNXHFRxcArrU0ZJSb9BIL56vmUt5FhlE'
# google link 'https://drive.google.com/u/0/uc?id=1NNXHFRxcArrU0ZJSb9BIL56vmUt5FhlE&export=download'
url = 'https://drive.google.com/uc?id=' + assistment_2009_download_id
import gdown, pandas as pd
!mkdir pykt-toolkit/data/assist2009
output = 'pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv'
gdown.download(url, output, quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1NNXHFRxcArrU0ZJSb9BIL56vmUt5FhlE
To: /content/pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv
100%|██████████| 64.4M/64.4M [00:00<00:00, 103MB/s]


'pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv'

In [ ]:
#assistment 2009 - open using windows-1252 encoding and save to appropriate place

df = pd.read_csv(
    "pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv",
    encoding='windows-1252'
  )
df_sample = df#.head(500) # to have a faster execution, comment out this line to use whole dataset
df_sample.to_csv("pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv")
df_sample.head()

<ipython-input-4-e2e900cb29bf>:3: DtypeWarning: Columns (17,18) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(


Unnamed: 0  order_id  assignment_id  user_id  assistment_id  problem_id  \
0           1  33022537         277618    64525          33139       51424   
1           2  33022709         277618    64525          33150       51435   
2           3  35450204         220674    70363          33159       51444   
3           4  35450295         220674    70363          33110       51395   
4           5  35450311         220674    70363          33196       51481   

   original  correct  attempt_count  ms_first_response  ... hint_count  \
0         1        1              1              32454  ...          0   
1         1        1              1               4922  ...          0   
2         1        0              2              25390  ...          0   
3         1        1              1               4859  ...          0   
4         1        0             14              19813  ...          3   

  hint_total  overlap_time  template_id  answer_id answer_text  first_action  \
0          3         32454        30799        NaN          26             0   
1          3          4922        30799        NaN          55             0   
2          3         42000        30799        NaN          88             0   
3          3          4859        30059        NaN          41             0   
4          4        124564        30060        NaN          65             0   

  bottom_hint opportunity  opportunity_original  
0         NaN           1                   1.0  
1         NaN           2                   2.0  
2         NaN           1                   1.0  
3         NaN           2                   2.0  
4         0.0           3                   3.0  

[5 rows x 31 columns]

In [ ]:
print(f'number of interactions: {len(df)}')
print(f"number of questions: {len(df['problem_id'].unique())}")
print(f"number of students: {len(df['user_id'].unique())}")

number of interactions: 346860
number of questions: 26688
number of students: 4217


Make sure, I use the same data as pyKT:


https://pykt-toolkit.readthedocs.io/en/latest/datasets.html
This dataset is made up of math exercises, collected from the free online tutoring ASSISTments platform in the school year 2009-2010. The dataset consists of **346,860 interactions, 4,217 students, and 26,688 questions** and is widely used and has been the standard benchmark for KT methods over the last decade.

# Pre-process, train, and eval - ASSIST2009

In [ ]:
!cd pykt-toolkit/examples/ ; python data_preprocess.py --dataset_name=assist2009

Namespace(dataset_name='assist2009', file_path='../data/peiyou/grade3_students_b_200.csv', min_seq_len=3, maxlen=200, kfold=5)
Start preprocessing data: assist2009
original interaction num: 346860, user num: 4217, question num: 26688, concept num: 125, avg(ins) per s: 82.2528, avg(c) per q: 1.2265, na: 8937
after drop interaction num: 283105, user num: 4163, question num: 17751, concept num: 125, avg(ins) per s: 68.005, avg(c) per q: 1.2265, na: 0
/usr/local/lib/python3.10/dist-packages/pykt/preprocess/assist2009_preprocess.py:25: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.
  for ui in ui_df:
346860,4217,26688,125,82.2528,1.2265,8937
283105,4163,17751,125,68.005,1.2265,0
--------------------------------------------------
rm: cannot remove '../data/assist2009/*.pkl': No such file or directory
delete bad stu num of 

# SAINT

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_saint_train.py --dataset_name=assist2009 --use_wandb=0 > saint_train.txt

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_predict.py --use_wandb=0 --save_dir='saved_model/assist2009_saint_qid_saved_model_42_0_0.2_256_0.001_8_4_0_1'

Namespace(bz=256, save_dir='saved_model/assist2009_saint_qid_saved_model_42_0_0.2_256_0.001_8_4_0_1', fusion_type='early_fusion,late_fusion', use_wandb=0)
model_name is saint
Read data from processed file: ../data/assist2009/test_sequences.csv_-1.pkl
file path: ../data/assist2009/test_sequences.csv, qlen: 904, clen: 904, rlen: 904
Start preprocessing ../data/assist2009/test_window_sequences.csv fold: _-1...
interaction_num: 62810
file path: ../data/assist2009/test_window_sequences.csv, qlen: 20362, clen: 20362, rlen: 20362
Start preprocessing ../data/assist2009/test_question_sequences.csv fold: _-1...
interaction_num: 61799
file path: ../data/assist2009/test_question_sequences.csv, qlen: 18292, clen: 18292, rlen: 18292
Start preprocessing ../data/assist2009/test_question_window_sequences.csv fold: _-1...
interaction_num: 61882
file path: ../data/assist2009/test_question_window_sequences.csv, qlen: 30610, clen: 30610, rlen: 30610
has test_question_file!
Start predicting model: saint, em

# DKVMN

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_dkvmn_train.py --dataset_name=assist2009 --use_wandb=0 > dkvmn_train.txt

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_predict.py --use_wandb=0 --save_dir='saved_model/assist2009_dkvmn_qid_saved_model_42_0_0.2_200_0.001_50_0_1'

Namespace(bz=256, save_dir='saved_model/assist2009_dkvmn_qid_saved_model_42_0_0.2_200_0.001_50_0_1', fusion_type='early_fusion,late_fusion', use_wandb=0)
model_name is dkvmn
Read data from processed file: ../data/assist2009/test_sequences.csv_-1.pkl
file path: ../data/assist2009/test_sequences.csv, qlen: 904, clen: 904, rlen: 904
Read data from processed file: ../data/assist2009/test_window_sequences.csv_-1.pkl
file path: ../data/assist2009/test_window_sequences.csv, qlen: 20362, clen: 20362, rlen: 20362
Read data from processed file: ../data/assist2009/test_question_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_sequences.csv, qlen: 18292, clen: 18292, rlen: 18292
Read data from processed file: ../data/assist2009/test_question_window_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_window_sequences.csv, qlen: 30610, clen: 30610, rlen: 30610
has test_question_file!
Start predicting model: dkvmn, embtype: qid, save_dir: saved_model/assist200